### PR TITLE
Refactor the layouters

### LINKED ISSUE #143: fitting predicates never get used with Unbounded PageWidth
Since https://github.com/quchen/prettyprinter/commit/ca702667ee90c38d2d4ac6d8a57838894ac36c5e we never use a `FittingPredicate` with `Unbounded` – the `PageWidth` is always an `AvailablePerLine n r`.

To make the code clearer, and potentially help performance, the page width used in the fitting predicates should have only a single constructor then.

Note that, as pointed out in https://github.com/quchen/prettyprinter/issues/142, we only actually use the "page width" `Int`, but not the "ribbon width" `Double` from `AvailablePerLine`.

### PR DESCRIPTION
**Pull request by sjakobi**
Closes #143.

### CODE REVIEW

**Comment by sjakobi** [comment]
I plan on trying to pass in `selectNicer` to `layoutWadlerLeijen` instead of a `FittingPredicate`. That should reduce some complexity e.g. for `layoutPretty`.

**Comment by sjakobi** [comment]
> I plan on trying to pass in `selectNicer` to `layoutWadlerLeijen` instead of a `FittingPredicate`.

One downside of passing in a function of type `Int -> Int -> Doc ann -> Doc ann -> Doc ann` would be that it's not obvious that the resulting `Doc` is one of the inputs. Maybe that's not too bad though, at least as long as `layoutWadlerLeijen` is purely internal…

**Comment by sjakobi** [comment]
This didn't quite go where I thought it would, but I think it's still a nice improvement.

Reviews are welcome! :)

File: prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs [hunk 1]
```diff
@@ -1731,8 +1744,14 @@ layoutPretty
...
-layoutPretty = layoutWadlerLeijen
-    (FittingPredicate (\_pWidth _minNestingLevel maxWidth sdoc -> fits maxWidth sdoc))
+layoutPretty (LayoutOptions pageWidth_@(AvailablePerLine lineLength ribbonFraction)) =
+    layoutWadlerLeijen
+        (FittingPredicate
+             (\lineIndent currentColumn _initialIndentY sdoc ->
+                 fits
+                     (remainingWidth lineLength ribbonFraction lineIndent currentColumn)
+                     sdoc))
+        pageWidth_
```
Comment (quchen) [inline prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: What’s this underscore for? Looks alien. Name conflict?

File: prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs [hunk 2]
```diff
@@ -1820,31 +1842,70 @@ layoutSmart = layoutWadlerLeijen (FittingPredicate fits)
...
+    layoutWadlerLeijen
+        (FittingPredicate
+            (\_lineIndent _currentColumn _initialIndentY sdoc -> not (failsOnFirstLine sdoc)))
+        Unbounded
+  where
+    -- See the Note [Detecting failure with Unbounded page width].
+    failsOnFirstLine :: SimpleDocStream ann -> Bool
+    failsOnFirstLine = go
+      where
+        go sds = case sds of
```
Comment (quchen) [inline prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: Unnecessary »go« wrapper, can be directly recursive

File: prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs (hunk 1 above)
Comment (sjakobi) [reply prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs]: `pageWidth` is the API name for the `WithPageWidth` constructor. Previously the variable was often called `pWidth` which I found hard to understand.

**Pull request merged**